File: OpsT360/src/app/pages/rfid-seal-simulator/services/roboflow-validation.service.ts
import { Injectable } from '@angular/core';

export type RoboflowPrediction = {
  x: number;
  y: number;
  width: number;
  height: number;
  class: string;
  confidence?: number;
};

type RoboflowPredictionBlock = {
  predictions?: RoboflowPrediction[];
};

type RoboflowWorkflowOutput = {
  predictions?: RoboflowPredictionBlock | RoboflowPrediction[];
  output_image?: {
    type?: string;
    value?: string;
  };
};

type RoboflowWorkflowResponse = {
  outputs?: RoboflowWorkflowOutput[];
};

type BoxTuple = [number, number, number, number];

export type ValidationBox = {
  box: BoxTuple;
  area: number;
};

export type ValidationEvidenceResult = {
  detectedLabels: ValidationBox[];
  isSuccessful: boolean;
  validatedImage: string | null;
  outputImage: string | null;
};

@Injectable({ providedIn: 'root' })
export class RoboflowValidationService {
  private readonly blueRatioThreshold = 0.3;
  private readonly areaSimilarityThreshold = 0.4;
  private readonly panelMin = 0.4;
  private readonly panelMax = 0.6;
  private readonly panelOverlapTolerance = 0.03;
  private readonly panelShrinkPx = 2;

  processResponse(response: unknown): ValidationBox[] {
    const predictions = this.extractPredictions(response);
    return predictions.map((prediction) => this.buildBox(prediction));
  }

  async isValidationSuccessful(response: unknown, baseImage?: string): Promise<boolean> {
    const result = await this.analyzeValidation(response, baseImage);
    return result.isSuccessful;
  }

  async analyzeValidation(response: unknown, baseImage?: string): Promise<ValidationEvidenceResult> {
    const normalizedBaseImage = this.normalizeBaseImage(baseImage);
    const predictions = this.extractPredictions(response);
    if (!predictions.length || !normalizedBaseImage) {
      return {
        detectedLabels: [],
        isSuccessful: false,
        validatedImage: null,
        outputImage: this.extractOutputImage(response),
      };
    }

    const imageElement = await this.loadImage(normalizedBaseImage);
    const canvas = document.createElement('canvas');
    canvas.width = imageElement.width;
    canvas.height = imageElement.height;
    const ctx = canvas.getContext('2d');
    if (!ctx) {
      return {
        detectedLabels: [],
        isSuccessful: false,
        validatedImage: null,
        outputImage: this.extractOutputImage(response),
      };
    }
    ctx.drawImage(imageElement, 0, 0);
    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);

    const validLabelBoxes = predictions
      .filter((prediction) => prediction.class === 'stickerOK')
      .map((prediction) => this.buildBox(prediction))
      .filter((box) => this.isBlueLabel(imageData, box.box));

    const finalBoxes = this.reduceLabelBoxes(validLabelBoxes);
    const accessPanels = predictions
      .filter((prediction) => prediction.class === 'access_panel')
      .map((prediction) => ({
        box: this.buildBox(prediction).box,
        confidence: prediction.confidence ?? 0,
      }));
    const mainPanel = accessPanels.length
      ? accessPanels.reduce((best, current) =>
        current.confidence > best.confidence ? current : best,
      )
      : null;

    if (!mainPanel || !finalBoxes.length) {
      return {
        detectedLabels: finalBoxes,
        isSuccessful: false,
        validatedImage: await this.buildValidatedImage(response, normalizedBaseImage),
        outputImage: this.extractOutputImage(response),
      };
    }

    const panelBox = this.shrinkBox(mainPanel.box, this.panelShrinkPx);
    const lowerBound = this.panelMin - this.panelOverlapTolerance;
    const upperBound = this.panelMax + this.panelOverlapTolerance;

    const isSuccessful = finalBoxes.every((box) => {
      const panelOverlap = this.intersectionArea(box.box, panelBox);
      const panelRatio = box.area ? panelOverlap / box.area : 0;
      return panelRatio >= lowerBound && panelRatio <= upperBound;
    });

    return {
      detectedLabels: finalBoxes,
      isSuccessful,
      validatedImage: await this.buildValidatedImage(response, normalizedBaseImage),
      outputImage: this.extractOutputImage(response),
    };
  }

  extractOutputImage(response: unknown): string | null {
    if (!response || typeof response !== 'object') {
      return null;
    }
    const responseObject = response as RoboflowWorkflowResponse;
    const outputs = responseObject.outputs ?? (Array.isArray(response) ? response : [response]);
    const outputImage = this.findOutputImage(outputs);
    if (!outputImage?.value) {
      return null;
    }
    return this.normalizeOutputImage(outputImage);
  }

  async buildValidatedImage(response: unknown, baseImage?: string): Promise<string | null> {
    const outputImage = this.extractOutputImage(response);
    const normalizedBaseImage = this.normalizeBaseImage(baseImage) ?? outputImage;
    if (!normalizedBaseImage) {
      return null;
    }

    const predictions = this.extractPredictions(response);
    if (!predictions.length) {
      return normalizedBaseImage;
    }

    const image = await this.loadImage(normalizedBaseImage);
    const canvas = document.createElement('canvas');
    canvas.width = image.width;
    canvas.height = image.height;
    const ctx = canvas.getContext('2d');
    if (!ctx) {
      return outputImage;
    }
    ctx.drawImage(image, 0, 0);
    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);

    const validLabelBoxes = predictions
      .filter((prediction) => prediction.class === 'stickerOK')
      .map((prediction) => this.buildBox(prediction))
      .filter((box) => this.isBlueLabel(imageData, box.box));

    const finalBoxes = this.reduceLabelBoxes(validLabelBoxes);
    const accessPanels = predictions
      .filter((prediction) => prediction.class === 'access_panel')
      .map((prediction) => ({
        box: this.buildBox(prediction).box,
        confidence: prediction.confidence ?? 0,
      }));
    const mainPanel = accessPanels.length
      ? accessPanels.reduce((best, current) =>
        current.confidence > best.confidence ? current : best,
      )
      : null;

    if (mainPanel) {
      for (const panel of accessPanels) {
        this.drawBox(ctx, panel.box, '#2563eb', 'access_panel', panel.box[0], panel.box[1]);
      }
    }

    if (mainPanel && finalBoxes.length) {
      const panelBox = this.shrinkBox(mainPanel.box, this.panelShrinkPx);
      const lowerBound = this.panelMin - this.panelOverlapTolerance;
      const upperBound = this.panelMax + this.panelOverlapTolerance;

      for (const box of finalBoxes) {
        const panelOverlap = this.intersectionArea(box.box, panelBox);
        const panelRatio = box.area ? panelOverlap / box.area : 0;
        const isValid = panelRatio >= lowerBound && panelRatio <= upperBound;
        const label = `${isValid ? 'OK' : 'BAD'} (${Math.round(panelRatio * 100)}% panel)`;
        this.drawBox(
          ctx,
          box.box,
          isValid ? '#22c55e' : '#ef4444',
          `stickerOK ${label}`,
          box.box[0],
          box.box[3] + 20,
        );
      }
    }

    return canvas.toDataURL('image/jpeg', 0.92);
  }

  private findOutputImage(outputs: unknown[]): { type?: string; value?: string } | null {
    for (const output of outputs) {
      const candidate = output as RoboflowWorkflowOutput;
      if (candidate?.output_image?.value) {
        return candidate.output_image;
      }
    }
    return null;
  }

  private normalizeOutputImage(outputImage: { type?: string; value?: string }): string | null {
    if (!outputImage.value) {
      return null;
    }
    if (outputImage.value.startsWith('data:')) {
      return outputImage.value;
    }
    if (outputImage.type === 'base64' || this.looksLikeBase64(outputImage.value)) {
      return `data:image/jpeg;base64,${outputImage.value}`;
    }
    return outputImage.value;
  }

  private normalizeBaseImage(baseImage?: string): string | null {
    if (!baseImage) {
      return null;
    }
    if (baseImage.startsWith('data:')) {
      return baseImage;
    }
    if (this.looksLikeBase64(baseImage)) {
      return `data:image/jpeg;base64,${baseImage}`;
    }
    return baseImage;
  }

  private looksLikeBase64(value: string): boolean {
    if (!value || value.length < 32) {
      return false;
    }
    return /^[A-Za-z0-9+/]+={0,2}$/.test(value);
  }

  private extractPredictions(response: unknown): RoboflowPrediction[] {
    if (!response || typeof response !== 'object') {
      return [];
    }
    const responseObject = response as RoboflowWorkflowResponse;
    const outputs = responseObject.outputs ?? (Array.isArray(response) ? response : [response]);
    for (const output of outputs) {
      const candidate = output as RoboflowWorkflowOutput;
      const predictionsBlock = candidate?.predictions;

      if (Array.isArray(predictionsBlock)) {
        return predictionsBlock;
      }

      if (predictionsBlock && typeof predictionsBlock === 'object') {
        return (predictionsBlock as RoboflowPredictionBlock).predictions ?? [];
      }
    }

    return [];
  }

  private buildBox(prediction: RoboflowPrediction): ValidationBox {
    const x1 = Math.round(prediction.x - prediction.width / 2);
    const y1 = Math.round(prediction.y - prediction.height / 2);
    const x2 = Math.round(prediction.x + prediction.width / 2);
    const y2 = Math.round(prediction.y + prediction.height / 2);

    return {
      box: [x1, y1, x2, y2],
      area: Math.max(0, x2 - x1) * Math.max(0, y2 - y1),
    };
  }

  private reduceLabelBoxes(validBoxes: ValidationBox[]): ValidationBox[] {
    if (validBoxes.length <= 1) {
      return validBoxes;
    }
    if (validBoxes.length === 2) {
      const [first, second] = validBoxes;
      const minArea = Math.min(first.area, second.area);
      const maxArea = Math.max(first.area, second.area);
      if (minArea / maxArea >= this.areaSimilarityThreshold) {
        return validBoxes;
      }
      return [first.area >= second.area ? first : second];
    }

    return [...validBoxes].sort((a, b) => b.area - a.area).slice(0, 2);
  }

  private isBlueLabel(imageData: ImageData, box: BoxTuple): boolean {
    const [x1, y1, x2, y2] = box;
    const startX = Math.max(0, x1);
    const startY = Math.max(0, y1);
    const endX = Math.min(imageData.width, x2);
    const endY = Math.min(imageData.height, y2);

    if (endX <= startX || endY <= startY) {
      return false;
    }

    let bluePixels = 0;
    let totalPixels = 0;

    const data = imageData.data;

    for (let y = startY; y < endY; y += 1) {
      for (let x = startX; x < endX; x += 1) {
        const idx = (y * imageData.width + x) * 4;
        const r = data[idx];
        const g = data[idx + 1];
        const b = data[idx + 2];

        const { h, s, v } = this.rgbToHsv(r, g, b);
        const hueOpenCv = h / 2;

        totalPixels += 1;
        // Python reference uses OpenCV HSV scale (H: 0-179), so 95-140 in cv2 == 190-280 in 0-360.
        if (hueOpenCv >= 95 && hueOpenCv <= 140 && s >= 0.235 && v >= 0.196) {
          bluePixels += 1;
        }
      }
    }

    if (!totalPixels) {
      return false;
    }

    return bluePixels / totalPixels >= this.blueRatioThreshold;
  }

  private rgbToHsv(r: number, g: number, b: number): { h: number; s: number; v: number } {
    const rNorm = r / 255;
    const gNorm = g / 255;
    const bNorm = b / 255;

    const max = Math.max(rNorm, gNorm, bNorm);
    const min = Math.min(rNorm, gNorm, bNorm);
    const delta = max - min;

    let h = 0;

    if (delta !== 0) {
      if (max === rNorm) {
        h = ((gNorm - bNorm) / delta) % 6;
      } else if (max === gNorm) {
        h = (bNorm - rNorm) / delta + 2;
      } else {
        h = (rNorm - gNorm) / delta + 4;
      }

      h *= 60;
      if (h < 0) {
        h += 360;
      }
    }

    const s = max === 0 ? 0 : delta / max;
    const v = max;

    return { h, s, v };
  }

  private intersectionArea(boxA: BoxTuple, boxB: BoxTuple): number {
    const x1 = Math.max(boxA[0], boxB[0]);
    const y1 = Math.max(boxA[1], boxB[1]);
    const x2 = Math.min(boxA[2], boxB[2]);
    const y2 = Math.min(boxA[3], boxB[3]);
    if (x2 <= x1 || y2 <= y1) {
      return 0;
    }
    return (x2 - x1) * (y2 - y1);
  }
  private shrinkBox(box: BoxTuple, pixels: number): BoxTuple {
    return [
      box[0] + pixels,
      box[1] + pixels,
      box[2] - pixels,
      box[3] - pixels,
    ];
  }

  private drawBox(
    ctx: CanvasRenderingContext2D,
    box: BoxTuple,
    color: string,
    label: string,
    labelX: number,
    labelY: number,
  ): void {
    const [x1, y1, x2, y2] = box;
    ctx.strokeStyle = color;
    ctx.lineWidth = 8;
    ctx.strokeRect(x1, y1, x2 - x1, y2 - y1);
    ctx.font = '16px sans-serif';
    ctx.fillStyle = color;
    ctx.fillText(label, labelX, Math.max(labelY, 20));
  }

  private loadImage(src: string): Promise<HTMLImageElement> {
    return new Promise((resolve, reject) => {
      const img = new Image();
      img.onload = () => resolve(img);
      img.onerror = () => reject(new Error('Failed to load image'));
      img.src = src;
    });
  }

}

File: OpsT360/Services/RoboflowValidationService.cs
using System.Text.Json.Nodes;
using SkiaSharp;

namespace OpsT360.Services;

public sealed record ValidationBox(int X1, int Y1, int X2, int Y2, double Area);

public sealed record RoboflowValidationResult(
    bool IsSuccessful,
    string? ValidatedImageBase64,
    string? OutputImageBase64,
    IReadOnlyList<ValidationBox> DetectedLabels);

public sealed class RoboflowValidationService
{
    private const double BlueRatioThreshold = 0.3;
    private const double AreaSimilarityThreshold = 0.4;
    private const double PanelMin = 0.4;
    private const double PanelMax = 0.6;
    private const double PanelOverlapTolerance = 0.03;
    private const int PanelShrinkPx = 2;

    public RoboflowValidationResult AnalyzeValidation(JsonNode? response, string? baseImageBase64)
    {
        var normalizedBaseImage = NormalizeBaseImage(baseImageBase64);
        var predictions = ExtractPredictions(response);
        var outputImage = ExtractOutputImage(response);

        if (!predictions.Any() || string.IsNullOrWhiteSpace(normalizedBaseImage))
        {
            return new RoboflowValidationResult(false, null, outputImage, Array.Empty<ValidationBox>());
        }

        using var bitmap = DecodeBitmap(normalizedBaseImage);
        if (bitmap is null)
        {
            return new RoboflowValidationResult(false, null, outputImage, Array.Empty<ValidationBox>());
        }

        var validLabelBoxes = predictions
            .Where(p => string.Equals(p.Class, "stickerOK", StringComparison.OrdinalIgnoreCase))
            .Select(BuildBox)
            .Where(box => IsBlueLabel(bitmap, box))
            .ToList();

        var finalBoxes = ReduceLabelBoxes(validLabelBoxes);

        var accessPanels = predictions
            .Where(p => string.Equals(p.Class, "access_panel", StringComparison.OrdinalIgnoreCase))
            .Select(p => new { Box = BuildBox(p), Confidence = p.Confidence ?? 0d })
            .ToList();

        var mainPanel = accessPanels.Count == 0
            ? null
            : accessPanels.OrderByDescending(p => p.Confidence).First();

        var validatedImage = BuildValidatedImage(bitmap, accessPanels.Select(p => p.Box).ToList(), finalBoxes, mainPanel?.Box);

        if (mainPanel is null || finalBoxes.Count == 0)
        {
            return new RoboflowValidationResult(false, validatedImage, outputImage, finalBoxes);
        }

        var panelBox = ShrinkBox(mainPanel.Box, PanelShrinkPx);
        var lowerBound = PanelMin - PanelOverlapTolerance;
        var upperBound = PanelMax + PanelOverlapTolerance;

        var success = finalBoxes.All(box =>
        {
            var overlap = IntersectionArea(box, panelBox);
            var ratio = box.Area > 0 ? overlap / box.Area : 0;
            return ratio >= lowerBound && ratio <= upperBound;
        });

        return new RoboflowValidationResult(success, validatedImage, outputImage, finalBoxes);
    }

    private static string? BuildValidatedImage(SKBitmap source, List<ValidationBox> accessPanels, List<ValidationBox> labelBoxes, ValidationBox? mainPanel)
    {
        using var image = SKImage.FromBitmap(source);
        using var surface = SKSurface.Create(new SKImageInfo(source.Width, source.Height));
        var canvas = surface.Canvas;
        canvas.Clear(SKColors.White);
        canvas.DrawImage(image, 0, 0);

        foreach (var panel in accessPanels)
        {
            DrawBox(canvas, panel, SKColor.Parse("#2563eb"), "access_panel", panel.X1, panel.Y1);
        }

        if (mainPanel is not null && labelBoxes.Count > 0)
        {
            var panelBox = ShrinkBox(mainPanel, PanelShrinkPx);
            var lowerBound = PanelMin - PanelOverlapTolerance;
            var upperBound = PanelMax + PanelOverlapTolerance;

            foreach (var box in labelBoxes)
            {
                var overlap = IntersectionArea(box, panelBox);
                var ratio = box.Area > 0 ? overlap / box.Area : 0;
                var isValid = ratio >= lowerBound && ratio <= upperBound;
                var label = $"stickerOK {(isValid ? "OK" : "BAD")} ({Math.Round(ratio * 100)}% panel)";
                DrawBox(canvas, box, isValid ? SKColor.Parse("#22c55e") : SKColor.Parse("#ef4444"), label, box.X1, box.Y2 + 20);
            }
        }

        using var snapshot = surface.Snapshot();
        using var data = snapshot.Encode(SKEncodedImageFormat.Jpeg, 92);
        if (data is null)
            return null;

        return $"data:image/jpeg;base64,{Convert.ToBase64String(data.ToArray())}";
    }

    private static void DrawBox(SKCanvas canvas, ValidationBox box, SKColor color, string label, float labelX, float labelY)
    {
        using var stroke = new SKPaint
        {
            Color = color,
            IsStroke = true,
            StrokeWidth = 8,
            IsAntialias = true
        };

        using var text = new SKPaint
        {
            Color = color,
            TextSize = 24,
            IsAntialias = true
        };

        canvas.DrawRect(new SKRect(box.X1, box.Y1, box.X2, box.Y2), stroke);
        canvas.DrawText(label, labelX, Math.Max(labelY, 24), text);
    }

    private static bool IsBlueLabel(SKBitmap bitmap, ValidationBox box)
    {
        var startX = Math.Max(0, box.X1);
        var startY = Math.Max(0, box.Y1);
        var endX = Math.Min(bitmap.Width, box.X2);
        var endY = Math.Min(bitmap.Height, box.Y2);

        if (endX <= startX || endY <= startY)
            return false;

        var bluePixels = 0;
        var total = 0;

        for (var y = startY; y < endY; y++)
        {
            for (var x = startX; x < endX; x++)
            {
                var color = bitmap.GetPixel(x, y);
                var (h, s, v) = RgbToHsv(color.Red, color.Green, color.Blue);
                var hueOpenCv = h / 2.0;

                total++;
                if (hueOpenCv >= 95 && hueOpenCv <= 140 && s >= 0.235 && v >= 0.196)
                    bluePixels++;
            }
        }

        if (total == 0)
            return false;

        return (double)bluePixels / total >= BlueRatioThreshold;
    }

    private static (double H, double S, double V) RgbToHsv(byte r, byte g, byte b)
    {
        var rNorm = r / 255d;
        var gNorm = g / 255d;
        var bNorm = b / 255d;

        var max = Math.Max(rNorm, Math.Max(gNorm, bNorm));
        var min = Math.Min(rNorm, Math.Min(gNorm, bNorm));
        var delta = max - min;

        double h = 0;
        if (delta != 0)
        {
            if (max == rNorm)
                h = ((gNorm - bNorm) / delta) % 6;
            else if (max == gNorm)
                h = (bNorm - rNorm) / delta + 2;
            else
                h = (rNorm - gNorm) / delta + 4;

            h *= 60;
            if (h < 0)
                h += 360;
        }

        var s = max == 0 ? 0 : delta / max;
        var v = max;
        return (h, s, v);
    }

    private static SKBitmap? DecodeBitmap(string? source)
    {
        if (string.IsNullOrWhiteSpace(source))
            return null;

        try
        {
            byte[]? bytes = null;
            if (source.StartsWith("data:", StringComparison.OrdinalIgnoreCase))
            {
                var comma = source.IndexOf(',');
                if (comma > -1)
                    bytes = Convert.FromBase64String(source[(comma + 1)..]);
            }
            else if (LooksLikeBase64(source))
            {
                bytes = Convert.FromBase64String(source);
            }
            else if (File.Exists(source))
            {
                bytes = File.ReadAllBytes(source);
            }

            if (bytes is null || bytes.Length == 0)
                return null;

            return SKBitmap.Decode(bytes);
        }
        catch
        {
            return null;
        }
    }

    private static List<Prediction> ExtractPredictions(JsonNode? response)
    {
        if (response is null)
            return new List<Prediction>();

        foreach (var output in EnumerateOutputs(response))
        {
            var predictionsNode = output?["predictions"];
            if (predictionsNode is JsonArray directArray)
            {
                return directArray
                    .Select(ParsePrediction)
                    .Where(p => p is not null)
                    .Select(p => p!)
                    .ToList();
            }

            var nested = predictionsNode?["predictions"] as JsonArray;
            if (nested is not null)
            {
                return nested
                    .Select(ParsePrediction)
                    .Where(p => p is not null)
                    .Select(p => p!)
                    .ToList();
            }
        }

        return new List<Prediction>();
    }

    private static string? ExtractOutputImage(JsonNode? response)
    {
        if (response is null)
            return null;

        foreach (var output in EnumerateOutputs(response))
        {
            var imageObj = output?["output_image"];
            var value = imageObj?["value"]?.GetValue<string>();
            if (string.IsNullOrWhiteSpace(value))
                continue;

            var type = imageObj?["type"]?.GetValue<string>();
            return NormalizeOutputImage(value, type);
        }

        return null;
    }

    private static IEnumerable<JsonNode?> EnumerateOutputs(JsonNode response)
    {
        if (response["outputs"] is JsonArray outputsArray)
        {
            foreach (var item in outputsArray)
                yield return item;

            yield break;
        }

        if (response is JsonArray rootArray)
        {
            foreach (var item in rootArray)
                yield return item;

            yield break;
        }

        yield return response;
    }

    private static string? NormalizeOutputImage(string value, string? type)
    {
        if (string.IsNullOrWhiteSpace(value))
            return null;

        if (value.StartsWith("data:", StringComparison.OrdinalIgnoreCase))
            return value;

        if (string.Equals(type, "base64", StringComparison.OrdinalIgnoreCase) || LooksLikeBase64(value))
            return $"data:image/jpeg;base64,{value}";

        return value;
    }

    private static string? NormalizeBaseImage(string? baseImage)
    {
        if (string.IsNullOrWhiteSpace(baseImage))
            return null;

        if (baseImage.StartsWith("data:", StringComparison.OrdinalIgnoreCase))
            return baseImage;

        if (LooksLikeBase64(baseImage))
            return $"data:image/jpeg;base64,{baseImage}";

        return baseImage;
    }

    private static bool LooksLikeBase64(string value)
    {
        if (string.IsNullOrWhiteSpace(value) || value.Length < 32)
            return false;

        return value.All(c => char.IsLetterOrDigit(c) || c is '+' or '/' or '=');
    }

    private static Prediction? ParsePrediction(JsonNode? node)
    {
        if (node is null)
            return null;

        var x = node["x"]?.GetValue<double>() ?? 0;
        var y = node["y"]?.GetValue<double>() ?? 0;
        var width = node["width"]?.GetValue<double>() ?? 0;
        var height = node["height"]?.GetValue<double>() ?? 0;
        var @class = node["class"]?.GetValue<string>();
        var confidence = node["confidence"]?.GetValue<double?>();

        if (string.IsNullOrWhiteSpace(@class))
            return null;

        return new Prediction(x, y, width, height, @class, confidence);
    }

    private static ValidationBox BuildBox(Prediction prediction)
    {
        var x1 = (int)Math.Round(prediction.X - prediction.Width / 2d);
        var y1 = (int)Math.Round(prediction.Y - prediction.Height / 2d);
        var x2 = (int)Math.Round(prediction.X + prediction.Width / 2d);
        var y2 = (int)Math.Round(prediction.Y + prediction.Height / 2d);

        var area = Math.Max(0, x2 - x1) * Math.Max(0, y2 - y1);
        return new ValidationBox(x1, y1, x2, y2, area);
    }

    private static List<ValidationBox> ReduceLabelBoxes(List<ValidationBox> validBoxes)
    {
        if (validBoxes.Count <= 1)
            return validBoxes;

        if (validBoxes.Count == 2)
        {
            var first = validBoxes[0];
            var second = validBoxes[1];
            var minArea = Math.Min(first.Area, second.Area);
            var maxArea = Math.Max(first.Area, second.Area);
            if (maxArea <= 0)
                return new List<ValidationBox>();

            return minArea / maxArea >= AreaSimilarityThreshold
                ? validBoxes
                : new List<ValidationBox> { first.Area >= second.Area ? first : second };
        }

        return validBoxes.OrderByDescending(b => b.Area).Take(2).ToList();
    }

    private static ValidationBox ShrinkBox(ValidationBox box, int pixels) =>
        new(box.X1 + pixels, box.Y1 + pixels, box.X2 - pixels, box.Y2 - pixels, Math.Max(0, (box.X2 - box.X1 - 2 * pixels) * (box.Y2 - box.Y1 - 2 * pixels)));

    private static double IntersectionArea(ValidationBox a, ValidationBox b)
    {
        var x1 = Math.Max(a.X1, b.X1);
        var y1 = Math.Max(a.Y1, b.Y1);
        var x2 = Math.Min(a.X2, b.X2);
        var y2 = Math.Min(a.Y2, b.Y2);
        if (x2 <= x1 || y2 <= y1)
            return 0;

        return (x2 - x1) * (y2 - y1);
    }

    private sealed record Prediction(double X, double Y, double Width, double Height, string Class, double? Confidence);
}
